<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="GeneCodeR">
<title>Analysis of Spatial Transcriptomic data • GeneCodeR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis of Spatial Transcriptomic data">
<meta property="og:description" content="GeneCodeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">GeneCodeR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2022.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/analysis_of_spatial_transcriptomics.html">Analysis of Spatial Transcriptomic data</a>
    <a class="dropdown-item" href="../articles/base_validation.html">Base validation</a>
    <a class="dropdown-item" href="../articles/displacement_validation.html">Displacement validation</a>
    <a class="dropdown-item" href="../articles/rotation_validation.html">Rotation validation</a>
    <a class="dropdown-item" href="../articles/signal_validation.html">Signal validation</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analysis of Spatial Transcriptomic data</h1>
                        <h4 data-toc-skip class="author"><a href="mailto:help@askexplain.com" class="email">help@askexplain.com</a></h4>
            
            <h4 data-toc-skip class="date">2022-12-18</h4>
      
      
      <div class="d-none name"><code>analysis_of_spatial_transcriptomics.Rmd</code></div>
    </div>

    
    
<div class="section level3">
<h3 id="genecoder-analysis-of-spatial-transcriptomic-data">GeneCodeR analysis of Spatial Transcriptomic data<a class="anchor" aria-label="anchor" href="#genecoder-analysis-of-spatial-transcriptomic-data"></a>
</h3>
<p>Imaging tissue contains genes spatially organised by the
extracellular matrix, individual cell types and groups of cell
compartments. Using linear algebra and vectors, the expected patterns of
the histology images are learned (example patterns of (1): extracellular
matrix stained in eosin, and (2): nuclei of cells stained in
hematoxylin). Assigning each sample image to a learned pattern (by the
Generative Encoder algorithm) is similar to assigning an image weight to
a pattern vector. Accordingly, genes are also assigned weights to the
same set of expected image patterns. This encodes the information into
the same space.</p>
<p>The library that wraps around the Generative Encoder algorithm
(gcode) is called GeneCodeR and provides a more comprehensive interface
to load data from path, run gcode, extract relevant transformations, and
validate the learned patterns.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Main libraries for analysis</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GeneCodeR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gcode</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Main libraries for plotting</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">grid</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">gridExtra</span><span class="op">)</span></span></code></pre></div>
<p>Please replace these paths</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">main_path</span> <span class="op">&lt;-</span> <span class="st">"~/Documents/main_files/AskExplain/Q4_2022/gcode/"</span></span>
<span></span>
<span><span class="co"># Please replace these paths</span></span>
<span><span class="va">path_to_save</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">main_path</span>,<span class="st">"./temp_save_dir/"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="va">path_to_work</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">main_path</span>,<span class="st">"./temp_work_dir/"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p>Download data and format</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">path_to_save</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in dir.create(path_to_save): '/Users/dbanh_side/Documents/main_files/</span></span>
<span><span class="co">## AskExplain/Q4_2022/gcode/./temp_save_dir' already exists</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">path_to_work</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in dir.create(path_to_work): '/Users/dbanh_side/Documents/main_files/</span></span>
<span><span class="co">## AskExplain/Q4_2022/gcode/./temp_work_dir' already exists</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">path_to_work</span><span class="op">)</span></span>
<span><span class="fu">curl</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/curl/man/curl_download.html" class="external-link">curl_download</a></span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://prod-dcd-datasets-cache-zipfiles.s3.eu-west-1.amazonaws.com/29ntw7sh4r-5.zip"</span>, destfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">path_to_work</span>,<span class="st">"he_et_al.zip"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/unzip.html" class="external-link">unzip</a></span><span class="op">(</span>zipfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">path_to_work</span>,<span class="st">"he_et_al.zip"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>, exdir <span class="op">=</span> <span class="va">path_to_work</span><span class="op">)</span></span>
<span><span class="va">path_to_work</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">path_to_work</span>,<span class="st">"29ntw7sh4r-5"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="va">path_to_work</span><span class="op">)</span></span>
<span></span>
<span><span class="va">main_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path_to_work</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">new_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"BC"</span>, replacement <span class="op">=</span> <span class="st">"BT"</span>, x <span class="op">=</span> <span class="va">main_files</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.rename</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">main_files</span>, to <span class="op">=</span> <span class="va">new_files</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [31] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [46] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [61] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [76] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">##  [91] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [106] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [121] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [136] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [151] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [166] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [181] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [196] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [211] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [226] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [241] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [256] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">## [271] TRUE TRUE TRUE</span></span></code></pre>
<p>GeneCodeR takes in three input files from a generic Spatial
Transcriptomic dataset, per tissue slide:</p>
<p>1 - coordinates 2 - gene expression 3 - images</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Input list of files</span></span>
<span></span>
<span><span class="va">file_path_list</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_path_framework.html">extract_path_framework</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">file_path_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path_to_work</span>,pattern <span class="op">=</span> <span class="st">"*Coords.tsv.gz"</span>,full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path_to_work</span>,pattern <span class="op">=</span> <span class="st">"*csv.gz"</span>,full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">file_path_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path_to_work</span>,pattern <span class="op">=</span> <span class="st">"*stdata.tsv.gz"</span>,full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">file_path_list</span><span class="op">$</span><span class="va">pixel</span><span class="op">$</span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path_to_work</span>,pattern <span class="op">=</span> <span class="st">"*jpg"</span>,full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">train_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">*</span><span class="fl">0.9</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">test_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="va">train_ids</span><span class="op">]</span></span>
<span></span>
<span><span class="va">train_file_path_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">file_path_list</span>,<span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">X</span><span class="op">$</span><span class="va">path</span><span class="op">[</span><span class="va">train_ids</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">test_file_path_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">file_path_list</span>,<span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">X</span><span class="op">$</span><span class="va">path</span><span class="op">[</span><span class="op">-</span><span class="va">train_ids</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>The GeneCodeR configuration file takes in several parameters.</p>
<p>The first set of parameters is information to extract spots - such as
spot size, and any augmentations such as displacement of the
coordinates, or rotation of the image.</p>
<p>The second set of parameters is for after the gcode model is learned
(via GeneCodeR wrapper). This is to transfrom from one modality, to
another modality (for example, from image to gene expression).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up genecoder configuration parameters</span></span>
<span></span>
<span><span class="va">genecoder.config</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_config_framework.html">extract_config_framework</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>The meta information file takes in several parameters per coordinate,
gene expression and image.</p>
<p>The first set of parameters is to read in the file. Here, example
parameters such as the file type (for example, is it a .csv, .tsv, .jpg,
.png, market matrix file (.mtx)?), and other important parameters based
on file type (for example, does it contain a header, what is the
separator …).</p>
<p>Importantly, another set of parameters is the factor identifier in
the coordinate file - which columns contain the x-coordinates,
y-coordinates and the label identifier for the spots? For example, the
label identifiers, are given by the rownames indicated by a 0.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up meta information</span></span>
<span></span>
<span><span class="va">meta_info_list</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_meta_framework.html">extract_meta_framework</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">file_type</span> <span class="op">&lt;-</span> <span class="st">"tsv"</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">file_type</span> <span class="op">&lt;-</span> <span class="st">"csv"</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">file_type</span> <span class="op">&lt;-</span> <span class="st">"tsv"</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">pixel</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">file_type</span> <span class="op">&lt;-</span> <span class="st">"image"</span></span>
<span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">sep</span> <span class="op">&lt;-</span> <span class="st">"\t"</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">header</span> <span class="op">&lt;-</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">header</span> <span class="op">&lt;-</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">header</span> <span class="op">&lt;-</span> <span class="cn">T</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">sep</span> <span class="op">&lt;-</span> <span class="st">","</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">sep</span> <span class="op">&lt;-</span> <span class="st">"\t"</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">row.names</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">quote</span> <span class="op">&lt;-</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">quote</span> <span class="op">&lt;-</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">quote</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">row.names</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">$</span><span class="va">meta</span><span class="op">$</span><span class="va">row.names</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">factor_id</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">factor_id</span><span class="op">$</span><span class="va">coord_x</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">factor_id</span><span class="op">$</span><span class="va">coord_y</span> <span class="op">&lt;-</span> <span class="fl">2</span></span></code></pre></div>
<p>To run the analysis appropriately, a set of common genes are required
for all spots.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">file_path_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">X</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu">read_file</span><span class="op">(</span>path <span class="op">=</span> <span class="va">file_path_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">path</span><span class="op">[</span><span class="va">X</span><span class="op">]</span>, meta_info <span class="op">=</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">read_file</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1</span></span>
<span><span class="co">## [1] 2</span></span>
<span><span class="co">## [1] 3</span></span>
<span><span class="co">## [1] 4</span></span>
<span><span class="co">## [1] 5</span></span>
<span><span class="co">## [1] 6</span></span>
<span><span class="co">## [1] 7</span></span>
<span><span class="co">## [1] 8</span></span>
<span><span class="co">## [1] 9</span></span>
<span><span class="co">## [1] 10</span></span>
<span><span class="co">## [1] 11</span></span>
<span><span class="co">## [1] 12</span></span>
<span><span class="co">## [1] 13</span></span>
<span><span class="co">## [1] 14</span></span>
<span><span class="co">## [1] 15</span></span>
<span><span class="co">## [1] 16</span></span>
<span><span class="co">## [1] 17</span></span>
<span><span class="co">## [1] 18</span></span>
<span><span class="co">## [1] 19</span></span>
<span><span class="co">## [1] 20</span></span>
<span><span class="co">## [1] 21</span></span>
<span><span class="co">## [1] 22</span></span>
<span><span class="co">## [1] 23</span></span>
<span><span class="co">## [1] 24</span></span>
<span><span class="co">## [1] 25</span></span>
<span><span class="co">## [1] 26</span></span>
<span><span class="co">## [1] 27</span></span>
<span><span class="co">## [1] 28</span></span>
<span><span class="co">## [1] 29</span></span>
<span><span class="co">## [1] 30</span></span>
<span><span class="co">## [1] 31</span></span>
<span><span class="co">## [1] 32</span></span>
<span><span class="co">## [1] 33</span></span>
<span><span class="co">## [1] 34</span></span>
<span><span class="co">## [1] 35</span></span>
<span><span class="co">## [1] 36</span></span>
<span><span class="co">## [1] 37</span></span>
<span><span class="co">## [1] 38</span></span>
<span><span class="co">## [1] 39</span></span>
<span><span class="co">## [1] 40</span></span>
<span><span class="co">## [1] 41</span></span>
<span><span class="co">## [1] 42</span></span>
<span><span class="co">## [1] 43</span></span>
<span><span class="co">## [1] 44</span></span>
<span><span class="co">## [1] 45</span></span>
<span><span class="co">## [1] 46</span></span>
<span><span class="co">## [1] 47</span></span>
<span><span class="co">## [1] 48</span></span>
<span><span class="co">## [1] 49</span></span>
<span><span class="co">## [1] 50</span></span>
<span><span class="co">## [1] 51</span></span>
<span><span class="co">## [1] 52</span></span>
<span><span class="co">## [1] 53</span></span>
<span><span class="co">## [1] 54</span></span>
<span><span class="co">## [1] 55</span></span>
<span><span class="co">## [1] 56</span></span>
<span><span class="co">## [1] 57</span></span>
<span><span class="co">## [1] 58</span></span>
<span><span class="co">## [1] 59</span></span>
<span><span class="co">## [1] 60</span></span>
<span><span class="co">## [1] 61</span></span>
<span><span class="co">## [1] 62</span></span>
<span><span class="co">## [1] 63</span></span>
<span><span class="co">## [1] 64</span></span>
<span><span class="co">## [1] 65</span></span>
<span><span class="co">## [1] 66</span></span>
<span><span class="co">## [1] 67</span></span>
<span><span class="co">## [1] 68</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">factor</span><span class="op">$</span><span class="va">common_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html" class="external-link">Reduce</a></span><span class="op">(</span><span class="st">"intersect"</span>,<span class="va">common_genes</span><span class="op">)</span></span></code></pre></div>
<p>Coordinate labels also need to be extracted.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">factor</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">train_file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu">read_file</span><span class="op">(</span>path <span class="op">=</span> <span class="va">train_file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">[</span><span class="va">X</span><span class="op">]</span>, meta_info <span class="op">=</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Gene expression data is extracted using the path list, meta
information, and configuration parameters.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract gex data</span></span>
<span><span class="va">gex_data</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/prepare_gex.html">prepare_gex</a></span><span class="op">(</span>file_path_list <span class="op">=</span> <span class="va">train_file_path_list</span>,meta_info_list <span class="op">=</span> <span class="va">meta_info_list</span>,config <span class="op">=</span> <span class="va">genecoder.config</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Extracting gex"</span></span>
<span><span class="co">## [1] "Preparing gex      1"</span></span>
<span><span class="co">## [1] "Preparing gex      2"</span></span>
<span><span class="co">## [1] "Preparing gex      3"</span></span>
<span><span class="co">## [1] "Preparing gex      4"</span></span>
<span><span class="co">## [1] "Preparing gex      5"</span></span>
<span><span class="co">## [1] "Preparing gex      6"</span></span>
<span><span class="co">## [1] "Preparing gex      7"</span></span>
<span><span class="co">## [1] "Preparing gex      8"</span></span>
<span><span class="co">## [1] "Preparing gex      9"</span></span>
<span><span class="co">## [1] "Preparing gex      10"</span></span>
<span><span class="co">## [1] "Preparing gex      11"</span></span>
<span><span class="co">## [1] "Preparing gex      12"</span></span>
<span><span class="co">## [1] "Preparing gex      13"</span></span>
<span><span class="co">## [1] "Preparing gex      14"</span></span>
<span><span class="co">## [1] "Preparing gex      15"</span></span>
<span><span class="co">## [1] "Preparing gex      16"</span></span>
<span><span class="co">## [1] "Preparing gex      17"</span></span>
<span><span class="co">## [1] "Preparing gex      18"</span></span>
<span><span class="co">## [1] "Preparing gex      19"</span></span>
<span><span class="co">## [1] "Preparing gex      20"</span></span>
<span><span class="co">## [1] "Preparing gex      21"</span></span>
<span><span class="co">## [1] "Preparing gex      22"</span></span>
<span><span class="co">## [1] "Preparing gex      23"</span></span>
<span><span class="co">## [1] "Preparing gex      24"</span></span>
<span><span class="co">## [1] "Preparing gex      25"</span></span>
<span><span class="co">## [1] "Preparing gex      26"</span></span>
<span><span class="co">## [1] "Preparing gex      27"</span></span>
<span><span class="co">## [1] "Preparing gex      28"</span></span>
<span><span class="co">## [1] "Preparing gex      29"</span></span>
<span><span class="co">## [1] "Preparing gex      30"</span></span>
<span><span class="co">## [1] "Preparing gex      31"</span></span>
<span><span class="co">## [1] "Preparing gex      32"</span></span>
<span><span class="co">## [1] "Preparing gex      33"</span></span>
<span><span class="co">## [1] "Preparing gex      34"</span></span>
<span><span class="co">## [1] "Preparing gex      35"</span></span>
<span><span class="co">## [1] "Preparing gex      36"</span></span>
<span><span class="co">## [1] "Preparing gex      37"</span></span>
<span><span class="co">## [1] "Preparing gex      38"</span></span>
<span><span class="co">## [1] "Preparing gex      39"</span></span>
<span><span class="co">## [1] "Preparing gex      40"</span></span>
<span><span class="co">## [1] "Preparing gex      41"</span></span>
<span><span class="co">## [1] "Preparing gex      42"</span></span>
<span><span class="co">## [1] "Preparing gex      43"</span></span>
<span><span class="co">## [1] "Preparing gex      44"</span></span>
<span><span class="co">## [1] "Preparing gex      45"</span></span>
<span><span class="co">## [1] "Preparing gex      46"</span></span>
<span><span class="co">## [1] "Preparing gex      47"</span></span>
<span><span class="co">## [1] "Preparing gex      48"</span></span>
<span><span class="co">## [1] "Preparing gex      49"</span></span>
<span><span class="co">## [1] "Preparing gex      50"</span></span>
<span><span class="co">## [1] "Preparing gex      51"</span></span>
<span><span class="co">## [1] "Preparing gex      52"</span></span>
<span><span class="co">## [1] "Preparing gex      53"</span></span>
<span><span class="co">## [1] "Preparing gex      54"</span></span>
<span><span class="co">## [1] "Preparing gex      55"</span></span>
<span><span class="co">## [1] "Preparing gex      56"</span></span>
<span><span class="co">## [1] "Preparing gex      57"</span></span>
<span><span class="co">## [1] "Preparing gex      58"</span></span>
<span><span class="co">## [1] "Preparing gex      59"</span></span>
<span><span class="co">## [1] "Preparing gex      60"</span></span>
<span><span class="co">## [1] "Preparing gex      61"</span></span>
<span><span class="co">## [1] "Done preparation!"</span></span></code></pre>
<p>Importantly, labels for the meta information from the gene expression
data extraction is required for extracting relevant spots that contain
gene expression greater than background.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Store meta data</span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">factor</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">gex_data</span><span class="op">$</span><span class="va">labels</span></span>
<span></span>
<span></span>
<span><span class="co"># Extract spot data</span></span>
<span></span>
<span><span class="va">genecoder.config</span><span class="op">$</span><span class="va">extract_spots</span><span class="op">$</span><span class="va">window_size</span> <span class="op">&lt;-</span> <span class="fl">30</span></span>
<span></span>
<span><span class="va">spot_data</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/prepare_spot.html">prepare_spot</a></span><span class="op">(</span>file_path_list <span class="op">=</span> <span class="va">train_file_path_list</span>,meta_info_list <span class="op">=</span> <span class="va">meta_info_list</span>,config <span class="op">=</span> <span class="va">genecoder.config</span>, gex_data <span class="op">=</span> <span class="va">gex_data</span><span class="op">$</span><span class="va">gex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Extracting spots"</span></span>
<span><span class="co">## [1] "Preparing spot      1"</span></span>
<span><span class="co">## [1] "Preparing spot      2"</span></span>
<span><span class="co">## [1] "Preparing spot      3"</span></span>
<span><span class="co">## [1] "Preparing spot      4"</span></span>
<span><span class="co">## [1] "Preparing spot      5"</span></span>
<span><span class="co">## [1] "Preparing spot      6"</span></span>
<span><span class="co">## [1] "Preparing spot      7"</span></span>
<span><span class="co">## [1] "Preparing spot      8"</span></span>
<span><span class="co">## [1] "Preparing spot      9"</span></span>
<span><span class="co">## [1] "Preparing spot      10"</span></span>
<span><span class="co">## [1] "Preparing spot      11"</span></span>
<span><span class="co">## [1] "Preparing spot      12"</span></span>
<span><span class="co">## [1] "Preparing spot      13"</span></span>
<span><span class="co">## [1] "Preparing spot      14"</span></span>
<span><span class="co">## [1] "Preparing spot      15"</span></span>
<span><span class="co">## [1] "Preparing spot      16"</span></span>
<span><span class="co">## [1] "Preparing spot      17"</span></span>
<span><span class="co">## [1] "Preparing spot      18"</span></span>
<span><span class="co">## [1] "Preparing spot      19"</span></span>
<span><span class="co">## [1] "Preparing spot      20"</span></span>
<span><span class="co">## [1] "Preparing spot      21"</span></span>
<span><span class="co">## [1] "Preparing spot      22"</span></span>
<span><span class="co">## [1] "Preparing spot      23"</span></span>
<span><span class="co">## [1] "Preparing spot      24"</span></span>
<span><span class="co">## [1] "Preparing spot      25"</span></span>
<span><span class="co">## [1] "Preparing spot      26"</span></span>
<span><span class="co">## [1] "Preparing spot      27"</span></span>
<span><span class="co">## [1] "Preparing spot      28"</span></span>
<span><span class="co">## [1] "Preparing spot      29"</span></span>
<span><span class="co">## [1] "Preparing spot      30"</span></span>
<span><span class="co">## [1] "Preparing spot      31"</span></span>
<span><span class="co">## [1] "Preparing spot      32"</span></span>
<span><span class="co">## [1] "Preparing spot      33"</span></span>
<span><span class="co">## [1] "Preparing spot      34"</span></span>
<span><span class="co">## [1] "Preparing spot      35"</span></span>
<span><span class="co">## [1] "Preparing spot      36"</span></span>
<span><span class="co">## [1] "Preparing spot      37"</span></span>
<span><span class="co">## [1] "Preparing spot      38"</span></span>
<span><span class="co">## [1] "Preparing spot      39"</span></span>
<span><span class="co">## [1] "Preparing spot      40"</span></span>
<span><span class="co">## [1] "Preparing spot      41"</span></span>
<span><span class="co">## [1] "Preparing spot      42"</span></span>
<span><span class="co">## [1] "Preparing spot      43"</span></span>
<span><span class="co">## [1] "Preparing spot      44"</span></span>
<span><span class="co">## [1] "Preparing spot      45"</span></span>
<span><span class="co">## [1] "Preparing spot      46"</span></span>
<span><span class="co">## [1] "Preparing spot      47"</span></span>
<span><span class="co">## [1] "Preparing spot      48"</span></span>
<span><span class="co">## [1] "Preparing spot      49"</span></span>
<span><span class="co">## [1] "Preparing spot      50"</span></span>
<span><span class="co">## [1] "Preparing spot      51"</span></span>
<span><span class="co">## [1] "Preparing spot      52"</span></span>
<span><span class="co">## [1] "Preparing spot      53"</span></span>
<span><span class="co">## [1] "Preparing spot      54"</span></span>
<span><span class="co">## [1] "Preparing spot      55"</span></span>
<span><span class="co">## [1] "Preparing spot      56"</span></span>
<span><span class="co">## [1] "Preparing spot      57"</span></span>
<span><span class="co">## [1] "Preparing spot      58"</span></span>
<span><span class="co">## [1] "Preparing spot      59"</span></span>
<span><span class="co">## [1] "Preparing spot      60"</span></span>
<span><span class="co">## [1] "Preparing spot      61"</span></span>
<span><span class="co">## [1] "Done preparation!"</span></span></code></pre>
<p>Relevant frameworks are set up to begin learning the Generative
Encoder model.</p>
<p>Firstly, set up the data list, which stores the gene expression and
spot data.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up data</span></span>
<span></span>
<span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">spot_data</span><span class="op">$</span><span class="va">spot</span>,<span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">spot_data</span><span class="op">$</span><span class="va">gex</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The join framework is set up, creating the model structure that
learns (1): a similar function for “all” datasets, or (2): separate
functions across gene expression and image pixel spots.</p>
<p>Here, only the incode and the beta parameters are set up to learn the
individual patterns between genes and images. Given the samples are
extracted from the same tissue site (image and genes at the same spot),
the sample patterns are shared. The feature patterns and the sample
patterns (see first paragraph - genes assigned to expected image
patterns of hematoxylin and eosin stains) are mapped to a shared latent
space via the alpha and beta codes. For example, the use of image
patterns is analogous to pattern matching to the example image spot, the
latent codes un-rotate any example image spots until it best fits the
image pattern.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up join information</span></span>
<span></span>
<span><span class="va">join</span> <span class="op">&lt;-</span> <span class="fu">gcode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gcode/man/extract_join_framework.html" class="external-link">extract_join_framework</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">alpha_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">beta_sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">alpha_signal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">beta_signal</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">complete</span><span class="op">$</span><span class="va">code</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">join</span><span class="op">$</span><span class="va">covariance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu">gcode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gcode/man/extract_references_framework.html" class="external-link">extract_references_framework</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">reference</span><span class="op">$</span><span class="va">data_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>Set up configuration information - the dimensions of the encoding
space for the samples (i_dim), and the features (j_dim), as well as the
dimensions of the latent code (k_dim) are set up. The initialisation for
these parameters are set up via the irlba library using partial Singular
Value Decomposition.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up gcode config</span></span>
<span><span class="va">gcode.config</span> <span class="op">&lt;-</span> <span class="fu">gcode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gcode/man/extract_config.html" class="external-link">extract_config</a></span><span class="op">(</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">init</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>alpha_sample<span class="op">=</span><span class="st">"irlba"</span>,beta_sample<span class="op">=</span><span class="st">"irlba"</span><span class="op">)</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">i_dim</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">j_dim</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">dimension_reduction</span> <span class="op">&lt;-</span> <span class="cn">F</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">max_iter</span> <span class="op">&lt;-</span> <span class="fl">100</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">seed</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">tol</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">n.cores</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">learn_rate</span> <span class="op">&lt;-</span> <span class="fl">1e-1</span></span>
<span><span class="va">gcode.config</span><span class="op">$</span><span class="va">batch_size</span> <span class="op">&lt;-</span> <span class="fl">500</span></span></code></pre></div>
<p>Run the Generative Encoder model</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set up gcode model</span></span>
<span></span>
<span><span class="va">genecoder.model</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/learn_model.html">learn_model</a></span><span class="op">(</span>data_list <span class="op">=</span> <span class="va">data_list</span>, config <span class="op">=</span> <span class="va">gcode.config</span>, join <span class="op">=</span> <span class="va">join</span>, reference <span class="op">=</span> <span class="va">reference</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Initialising alpha_sample with : irlba"</span></span>
<span><span class="co">## [2] "Initialising beta_sample with : irlba"</span></span></code></pre>
<pre><code><span><span class="co">## Warning in irlba::irlba(as.matrix(x), nu = config$i_dim, maxit = 1): did not</span></span>
<span><span class="co">## converge--results might be invalid!; try increasing work or maxit</span></span></code></pre>
<pre><code><span><span class="co">## Warning in irlba::irlba(as.matrix(x), nv = config$j_dim, maxit = 1): did not</span></span>
<span><span class="co">## converge--results might be invalid!; try increasing work or maxit</span></span>
<span></span>
<span><span class="co">## Warning in irlba::irlba(as.matrix(x), nv = config$j_dim, maxit = 1): did not</span></span>
<span><span class="co">## converge--results might be invalid!; try increasing work or maxit</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Beginning gcode learning with:    Sample dimension reduction (config$i_dim): 500    Feature dimension reduction (config$j_dim): 500    Latent invariant dimension (config$k_dim):     Tolerance Threshold: 1   Maximum number of iterations: 100   Verbose: TRUE"</span></span>
<span><span class="co">## [1] "Iteration: 1 with Tolerance of: 21.1634805067891"</span></span>
<span><span class="co">## [1] "Iteration: 2 with Tolerance of: 63.4720802714958"</span></span>
<span><span class="co">## [1] "Iteration: 3 with Tolerance of: 8.75341078830188"</span></span>
<span><span class="co">## [1] "Iteration: 4 with Tolerance of: 82.3722229279707"</span></span>
<span><span class="co">## [1] "Iteration: 5 with Tolerance of: 0.272863785887901"</span></span>
<span><span class="co">## [1] "Learning has converged for gcode, beginning (if requested) dimension reduction"</span></span>
<span><span class="co">## [1] "Done! Total runtime of   43.044757048289"</span></span></code></pre>
<p>Now begins the testing validation phase…</p>
<p>First extract the test set of the coordinate factor labels.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Evaluate on test set</span></span>
<span></span>
<span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">factor</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">test_file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">)</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">X</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/row.names.html" class="external-link">row.names</a></span><span class="op">(</span><span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu">read_file</span><span class="op">(</span>path <span class="op">=</span> <span class="va">test_file_path_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">path</span><span class="op">[</span><span class="va">X</span><span class="op">]</span>, meta_info <span class="op">=</span> <span class="va">meta_info_list</span><span class="op">$</span><span class="va">coord</span><span class="op">$</span><span class="va">read_file</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>Next extract the test set of the gene expression data.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract test gex data</span></span>
<span><span class="va">test_gex_data</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/prepare_gex.html">prepare_gex</a></span><span class="op">(</span>file_path_list <span class="op">=</span> <span class="va">test_file_path_list</span>,meta_info_list <span class="op">=</span> <span class="va">meta_info_list</span>,config <span class="op">=</span> <span class="va">genecoder.config</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Extracting gex"</span></span>
<span><span class="co">## [1] "Preparing gex      1"</span></span>
<span><span class="co">## [1] "Preparing gex      2"</span></span>
<span><span class="co">## [1] "Preparing gex      3"</span></span>
<span><span class="co">## [1] "Preparing gex      4"</span></span>
<span><span class="co">## [1] "Preparing gex      5"</span></span>
<span><span class="co">## [1] "Preparing gex      6"</span></span>
<span><span class="co">## [1] "Preparing gex      7"</span></span>
<span><span class="co">## [1] "Done preparation!"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_info_list</span><span class="op">$</span><span class="va">gex</span><span class="op">$</span><span class="va">factor</span><span class="op">$</span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="va">test_gex_data</span><span class="op">$</span><span class="va">labels</span></span></code></pre></div>
<p>Save relevant information, including the model, the test meta
information, the test file list and the test gene expression data. This
is to analyse results later.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base_test_spot_data</span> <span class="op">&lt;-</span> <span class="fu">GeneCodeR</span><span class="fu">::</span><span class="fu"><a href="../reference/prepare_spot.html">prepare_spot</a></span><span class="op">(</span>file_path_list <span class="op">=</span> <span class="va">test_file_path_list</span>,meta_info_list <span class="op">=</span> <span class="va">meta_info_list</span>,config <span class="op">=</span> <span class="va">genecoder.config</span>, gex_data <span class="op">=</span> <span class="va">test_gex_data</span><span class="op">$</span><span class="va">gex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Extracting spots"</span></span>
<span><span class="co">## [1] "Preparing spot      1"</span></span>
<span><span class="co">## [1] "Preparing spot      2"</span></span>
<span><span class="co">## [1] "Preparing spot      3"</span></span>
<span><span class="co">## [1] "Preparing spot      4"</span></span>
<span><span class="co">## [1] "Preparing spot      5"</span></span>
<span><span class="co">## [1] "Preparing spot      6"</span></span>
<span><span class="co">## [1] "Preparing spot      7"</span></span>
<span><span class="co">## [1] "Done preparation!"</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">non_zero_markers</span> <span class="op">&lt;-</span> <span class="va">base_test_spot_data</span><span class="op">$</span><span class="va">gex</span><span class="op">&gt;</span><span class="fl">0</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save.image</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span>sep<span class="op">=</span><span class="st">""</span>,<span class="va">path_to_save</span>,<span class="st">"all_genecoder.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
